### DISCUSSION TITLE
Build liboqs and oqs-provider without installation

### DISCUSSION THREAD

**Discussion by beldmit** [open-quantum-safe/oqs-provider#300] [Q&A]
I'm trying to locally build openssl master, liboqs master, and oqs-provider

No problems with openssl
```
mkdir build
cd build
cmake -DOPENSSL_ROOT_DIR=~/upstream/openssl -DBUILD_SHARED_LIBS=ON ..
make
```
works perfectly

When I try to build oqs-provider (also from newly created `build` directory), it either links against my installed liboqs, or, when I uninstall it, fails to find liboqs though I tried to set `liboqs_DIR` variable to liboqs build directory, a subdir with liboqs.so and a subdir with liboqs cmake files.

What is the proper way to consequently build all the components in chain?

**Comment by ghost** [comment]
Hi @beldmit,

Do you actually install liboqs using `cmake --install build` (or `cd build && make install`)?

If so, a directory called `liboqs` should exist under `$PREFIX/lib/cmake`. It must be the value to `liboqs_DIR`:

```bash
$ cmake -B build -Dliboqs_DIR=$PREFIX/lib/cmake/liboqs -DOPENSSL_ROOT_DIR=$PREFIX/
$ cmake --build build/
```

**Reply by beldmit** [reply]
Again, I'd like to not install liboqs and build against the locally built version

**Reply by baentsch** [reply]
Re-reading this thread, isn't @thb-sb 's answer exactly what you need, @beldmit ? Just set liboqs_DIR to your pre-existing/installed `liboqs` install. For me, setting the env var `liboqs_DIR` to that directory does the trick on some arcane OSs, too (instead of/in addition to setting the `cmake` define).

**Reply by beldmit** [reply]
I'm sorry for the delay, I was unable to make experiments for some time and to check. I will return to this task later

**Reply by beldmit** [reply]
Looks like it doesn't work for me

`$ cmake -Dliboqs_DIR=/home/dbelyavs/work/hackathon/liboqs/build/src/ ..`

```
..
-- Found OpenSSL: /usr/lib64/libcrypto.so (found suitable version "3.1.1", minimum required is "3.0")  
CMake Error at /home/dbelyavs/work/hackathon/liboqs/build/src/liboqsConfig.cmake:29 (include):
  include could not find requested file:

    /home/dbelyavs/work/hackathon/liboqs/build/src/liboqsTargets.cmake
Call Stack (most recent call first):
  CMakeLists.txt:71 (find_package)


CMake Error at CMakeLists.txt:72 (get_target_property):
  get_target_property() called with non-existent target "OQS::oqs".
```

**Reply by baentsch** [reply]
The path passed to `liboqs_DIR` looks very fishy to me: All package files are in it, but in a completely different structure than `cmake` expects them (see [find_package](https://cmake.org/cmake/help/latest/command/find_package.html) documentation) when loading a required package, so the failure does not appear unexpected. I guess you don't like those scripts, but allow me to reiterate regardless that the way it's done in the build scripts is to install `liboqs` to a temp directory (and pass that to `liboqs_DIR`) such as to allow `cmake::find_package` to find all dependent files from `liboqs` (where it expects them). This may not be "elegant" but it works -- and in my eyes makes sense: `find_package` is meant to find _installed_ packages, not merely built ones. If you find another way, please let us know.

**Reply by beldmit** [reply]
Thank you very much!

**Comment by baentsch** [comment]
> 
> What is the proper way to consequently build all the components in chain?
> 

Did you check out 'scripts/fullbuild.sh'? Also the Dockerfiles of any integration, e.g 'curl' may be helpful.

**Reply by beldmit** [reply]
Thank you, will take a look
